### PR TITLE
i#6412: Separate raw2trace impl for sharing with tracer

### PR DESCRIPTION
**Pull request by abhinav92003**
Extracts raw2trace code that is shared with the drmemtrace tracer into
separate raw2trace_shared.h and raw2trace_shared.cpp. This is to allow
us to skip including the whole raw2trace implementation (and all the
libraries required by it like libipt) when building a separate library only for
tracing, like drmemtrace and drmemtrace_static.

This moves raw2trace's implementation of the drmemtrace.h
drmemtrace_get_timestamp_from_offline_trace, trace_metadata_reader_t
and various compile constants shared between raw2trace and the tracer.

drmemtrace and drmemtrace_static do not have any users of
drmemtrace_get_timestamp_from_offline_trace today, but some third-party
code does. Adding a test that links drmemtrace_static and tries to use
drmemtrace_get_timestamp_from_offline_trace from drmemtrace.h reveals
the missing impl in a build error. We add raw2trace_shared.cpp to
add_drmemtrace sources to fix this.

Also adds a CMake condition that checks that raw2trace.cpp should not be
part of add_drmemtrace sources. This is non-exhaustive of course as there
are many sources that are required only in offline raw trace conversion but
not during live tracing.

Unfortunately, our include_directories makes raw2trace.h still accessible
for inclusion to various tracer impl like tracer.cpp and output.cpp, so there's
no automated check on tracer code depending on header-only stuff in
raw2trace.h (or any other offline raw trace conversion only header). Added
a TODO for that.

Issue: #6412

### CODE REVIEW

**Comment by abhinav92003** [comment]
There seems to be a real consistent failure in x86-32-ubuntu22 from this PR. Looking.

**Comment by abhinav92003** [comment]
> There seems to be a real consistent failure in x86-32-ubuntu22 from this PR. Looking.

It's actually some recent flakiness. Xref #6416 and #6417.

File: clients/drcachesim/tracer/raw2trace.h [hunk 1]
```diff
@@ -83,23 +84,11 @@ namespace drmemtrace {
 #    define DEBUG_ASSERT(x) /* nothing */
 #endif
 
-#define OUTFILE_SUFFIX "raw"
 #ifdef BUILD_PT_POST_PROCESSOR
 #    define OUTFILE_SUFFIX_PT "raw.pt"
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.h:88]: This is part of the OUTFILE_SUFFIX* group though: it seems odd to separate it.  Should we move it to _shared.h too?

File: clients/drcachesim/tracer/raw2trace.h [hunk 2]
```diff
@@ -83,23 +84,11 @@ namespace drmemtrace {
...
-#endif
-#ifdef HAS_SNAPPY
-#    define OUTFILE_SUFFIX_SZ "raw.sz"
-#endif
-#ifdef HAS_LZ4
-#    define OUTFILE_SUFFIX_LZ4 "raw.lz4"
-#endif
-#define OUTFILE_SUBDIR "raw"
+
 #define WINDOW_SUBDIR_PREFIX "window"
```
Comment (derekbruening) [inline clients/drcachesim/tracer/raw2trace.h:91]: It also seems like this should live with the _FORMAT one: should we move them all as a group.

**Review by derekbruening — APPROVED** [review]

File: clients/drcachesim/CMakeLists.txt [hunk 3]
```diff
@@ -208,6 +208,7 @@ endif (BUILD_PT_POST_PROCESSOR)
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:208]: Done.

File: clients/drcachesim/CMakeLists.txt [hunk 4]
```diff
@@ -292,6 +293,13 @@ use_DynamoRIO_extension(drcachesim drutil_static)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/simulator)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/common)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/reader)
+# TODO i#6412: raw2trace.h and related headers that are used only in raw2trace
+# conversion are also a part of this directory. Ideally we'd like for online
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:297]: Done.

File: clients/drcachesim/CMakeLists.txt [hunk 5]
```diff
@@ -461,6 +473,20 @@ macro(add_drmemtrace name type)
...
   endif ()
+
+  # i#6412: All implementation shared between the online tracer and offline
+  # raw2trace should be in the separate raw2trace_shared files. This is so that
+  # we can keep the online tracing library smaller. The following is a
+  # non-exhaustive condition that flags when raw2trace implementation is
+  # accidentally added. Including raw2trace.cpp would necessitate inclusion of
+  # various other implementations (like for the trace reader_t), which should
+  # all be avoided here.
+  list (FIND drmemtrace_srcs "tracer/raw2trace.cpp" _raw2trace_index)
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:484]: Just to signify that it's a local not intended to be used elsewhere.

File: clients/drcachesim/CMakeLists.txt [hunk 6]
```diff
@@ -1079,6 +1105,19 @@ if (BUILD_TESTS)
     set_tests_properties(tool.drcacheoff.skip_unit_tests PROPERTIES
       TIMEOUT ${test_seconds})
   endif ()
+
+  add_executable(drmemtrace_no_raw2trace_tests
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:1109]: Done.

File: clients/drcachesim/tests/drmemtrace_no_raw2trace_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    offline_entry_t entry;
+    entry.timestamp.type = OFFLINE_TYPE_TIMESTAMP;
+    entry.timestamp.usec = value;
+    return entry;
+}
+
+bool
+test_get_timestamp()
+{
+    offline_entry_t bad_buffer[1] = { make_timestamp(456) };
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/tests/drmemtrace_no_raw2trace_tests.cpp:80]: Ah, that's a remnant from when I thought we need to "fix" is_thread_start to set the error string for the cases where it returns false, but later realized that it is intentional. Changed the identifier.

File: clients/drcachesim/tracer/raw2trace.h (hunk 2 above)
Comment (abhinav92003) [reply clients/drcachesim/tracer/raw2trace.h:91]: Done.

File: clients/drcachesim/tracer/raw2trace.h (hunk 1 above)
Comment (abhinav92003) [reply clients/drcachesim/tracer/raw2trace.h:88]: Done.

File: clients/drcachesim/CMakeLists.txt [hunk 8]
```diff
@@ -461,6 +473,20 @@ macro(add_drmemtrace name type)
       common/crc32c.cpp
       )
   endif ()
+
+  # i#6412: All implementation shared between the online tracer and offline
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:477]: Done.

File: clients/drcachesim/CMakeLists.txt [hunk 9]
```diff
@@ -346,7 +354,10 @@ install_client_nonDR_header(drmemtrace simulator/cache_simulator_create.h)
 install_client_nonDR_header(drmemtrace simulator/tlb_simulator_create.h)
 install_client_nonDR_header(drmemtrace tools/view_create.h)
 install_client_nonDR_header(drmemtrace tools/func_view_create.h)
+# TODO i#6412: Create a separate directory for non-tracer headers so that
+# we can more cleanly separate online tracer and raw2trace code.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:358]: Duplicated this and kept some of the above.

File: clients/drcachesim/CMakeLists.txt [hunk 10]
```diff
@@ -292,6 +293,13 @@ use_DynamoRIO_extension(drcachesim drutil_static)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/simulator)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/common)
 include_directories(${CMAKE_CURRENT_SOURCE_DIR}/reader)
+# TODO i#6412: raw2trace.h and related headers that are used only in raw2trace
+# conversion are also a part of this directory. Ideally we'd like for online
+# tracing code to not be able to include those headers. Today, raw2trace
+# implementation that is shared between the online tracer and raw2trace is
+# extracted into raw2trace_shared files, and raw2trace.cpp is disallowed in
+# drmemtrace and drmemtrace_static sources (a non-exhaustive check) to achieve
+# some separation.
```
[the remark this answers is not in the document]
Comment (abhinav92003) [reply clients/drcachesim/CMakeLists.txt:302]: The header check is missing yes, but also forbidding only raw2trace.cpp is not sufficient. A user may accidentally include some other raw2trace-only implementation (other than raw2trace.cpp).

**Pull request merged**